src/west/manifest.py

```diff
@@ -150,1504 +427,1302 @@ def _flags_ok(flags):
 
-    @staticmethod
-    def from_file(source_file=None, **kwargs):
-        '''Manifest object factory given a source YAML file.
+    Attributes:
 
-        The default behavior is to find the current west workspace's
-        manifest file and resolve it.
+    - ``name``: project's unique name
+    - ``url``: project fetch URL
+    - ``revision``: revision to fetch from ``url`` when the
+      project is updated
+    - ``path``: relative path to the project within the workspace
+      (i.e. from ``topdir`` if that is set)
+    - ``abspath``: absolute path to the project in the native path name
+      format (or ``None`` if ``topdir`` is)
+    - ``posixpath``: like ``abspath``, but with slashes (``/``) as
+      path separators
+    - ``clone_depth``: clone depth to fetch when first cloning the
+      project, or ``None`` (the revision should not be a SHA
+      if this is used)
+    - ``west_commands``: list of YAML files where extension commands in
+      the project are declared
+    - ``topdir``: the top level directory of the west workspace
+      the project is part of, or ``None``
+    - ``remote_name``: the name of the remote which should be set up
+      when the project is being cloned (default: 'origin')
+    '''
 
-        Results depend on the keyword arguments given in *kwargs*:
+    def __eq__(self, other):
+        return NotImplemented
 
-            - If both *source_file* and *topdir* are given, the
-              returned Manifest object is based on the data in
-              *source_file*, rooted at *topdir*. The configuration
-              files are not read in this case. This allows parsing a
-              manifest file "as if" its project hierarchy were rooted
-              at another location in the system.
+    def __repr__(self):
+        return (f'Project("{self.name}", "{self.url}", '
+                f'revision="{self.revision}", path={repr(self.path)}, '
+                f'clone_depth={self.clone_depth}, '
+                f'west_commands={self.west_commands}, '
+                f'topdir={repr(self.topdir)})')
 
-            - If neither *source_file* nor *topdir* is given, the file
-              system is searched for *topdir*. That workspace's
-              ``manifest.path`` configuration option is used to find
-              *source_file*, ``topdir/<manifest.path>/west.yml``.
+    def __str__(self):
+        path_repr = repr(self.abspath or self.path)
+        return f'<Project {self.name} ({path_repr}) at {self.revision}>'
 
-            - If only *source_file* is given, *topdir* is found
-              starting there. The directory containing *source_file*
-              doesn't have to be ``manifest.path`` in this case.
+    def __init__(self, name: str, url: str,
+                 revision: Optional[str] = None,
+                 path: Optional[str] = None,
+                 clone_depth: Optional[int] = None,
+                 west_commands: Optional[Union[str, List[str]]] = None,
```

Comment:
As `Optional[Union[str, List[str]]]` is used in quite a few places a custom type may be useful.